File: 2021/day04_1.py
# pylint: disable=missing-module-docstring, missing-function-docstring


from typing import Any

type Cell = tuple[int, bool]
type Row = list[Cell]
type Board = list[Row]


def solution(content: list[str]) -> Any:
    winning_board, winning_number = play(
        parse_boards("\n".join(content[2:]).split("\n\n")),
        [int(x) for x in content[0].split(",")],
    )
    return (
        sum(
            num
            for num, check in [cell for row in winning_board for cell in row]
            if not check
        )
        * winning_number
    )


def is_winning_line(line: list[Cell]):
    return all(check for _, check in line)


def transpose(board: Board):
    return [[row[index] for row in board] for index in range(len(board))]


def is_winner(board: Board):
    return any(is_winning_line(row) for row in board) or any(
        is_winning_line(col) for col in transpose(board)
    )


def winner(boards: list[Board]):
    for board in boards:
        if is_winner(board):
            return board
    return None


def play(boards: list[Board], numbers: list[int]):
    current = boards.copy()
    for number in numbers:
        current = [
            [[(num, check or num == number) for num, check in row] for row in board]
            for board in current
        ]

        if winning_board := winner(current):
            return (winning_board, number)

    return ([], 0)


def parse_boards(lines: list[str]):
    boards: list[Board] = [[]]
    for line in lines:
        if line:
            if not boards[-1]:
                boards[-1] = []
            boards[-1].append([(int(x), False) for x in line.split(" ") if x])
        else:
            boards.append([])
    return boards


def main() -> None:
    with open("day04.dat", "r", encoding="utf8") as file:
        print(solution(file.readlines()))


if __name__ == "__main__":
    main()

File: 2021/day04-1.ts
type Cell = { num: number; check: boolean };
type Board = Cell[][];

function solution(content: string) {
  const lines = content.split("\n");
  const numbers = lines[0].split(",").map(Number);
  const boards = parseBoards(lines.slice(2).join("\n").split("\n\n"));
  const [winningBoard, winningNumber] = play(boards, numbers);
  return (
    winningNumber *
    winningBoard
      .flat()
      .filter(({ check }) => !check)
      .map(({ num }) => num)
      .reduce((acc, cur) => acc + cur, 0)
  );
}

function isWinningLine(line: Cell[]) {
  return line.every(({ check }) => check);
}

function transpose(board: Board) {
  return board.map((_, index) => board.map((row) => row[index]));
}

function isWinner(board: Board) {
  return board.some(isWinningLine) || transpose(board).some(isWinningLine);
}

function winner(boards: Board[]) {
  for (const board of boards) {
    if (isWinner(board)) {
      return board;
    }
  }
  return null;
}

function play(boards: Board[], numbers: number[]): [Board, number] {
  let currentBoards = boards.slice();
  for (const anumber of numbers) {
    currentBoards = currentBoards.map((board) =>
      board.map((row) =>
        row.map(({ num, check }) => ({ num, check: check || num === anumber }))
      )
    );
    const winningBoard = winner(currentBoards);
    if (winningBoard) {
      return [winningBoard, anumber];
    }
  }
  return [[], 0];
}

function parseBoards(blocks: string[]) {
  return blocks.map((block) =>
    block.split("\n").map((line) =>
      line
        .split(" ")
        .filter((item) => item)
        .map((item) => ({ check: false, num: Number(item) }))
    )
  );
}

Deno.readTextFile("day04.dat")
  .then((content) => content.trim())
  .then(solution)
  .then(console.log);
